Plugins/NativeCamera/Source/NativeCameraAndroid.cpp

```diff
@@ -0,0 +1,411 @@
+#include <string>
+#include <camera/NdkCameraManager.h>
+#include <camera/NdkCameraCaptureSession.h>
+#include <camera/NdkCameraDevice.h>
+#include <camera/NdkCameraError.h>
+#include <camera/NdkCameraManager.h>
+#include <camera/NdkCameraMetadata.h>
+#include <camera/NdkCameraMetadataTags.h>
+#include <camera/NdkCameraWindowType.h>
+#include <camera/NdkCaptureRequest.h>
+#include <media/NdkImageReader.h>
+#include <android/native_window_jni.h>
+#include <AndroidExtensions/JavaWrappers.h>
+#include <AndroidExtensions/Globals.h>
+#include <android/log.h>
+#include <GLES2/gl2.h>
+#include <GLES2/gl2ext.h>
+#include <GLES3/gl3.h>
+#include <EGL/egl.h>
+#include <bgfx/bgfx.h>
+#include <bgfx/platform.h>
+#include <arcana/threading/task.h>
+#include <arcana/threading/dispatcher.h>
+
+using namespace android;
+using namespace android::global;
+
+namespace Babylon::Plugins::Internal
+{
+    const int PERMISSION_REQUEST_ID{ 8436 };
+
+    struct CameraInterfaceAndroid : public CameraInterface
+    {
+        CameraInterfaceAndroid(uint32_t width, uint32_t height, bool frontCamera);
+        virtual ~CameraInterfaceAndroid();
+        void UpdateCameraTexture(bgfx::TextureHandle textureHandle) override;
+
+        static constexpr char CAMERA_VERT_SHADER[]{ R"(#version 300 es
+            precision highp float;
+            out vec2 cameraFrameUV;
+            void main() {
+                cameraFrameUV = vec2(gl_VertexID&1, (gl_VertexID &2)>>1) * 2.f;
+                gl_Position = vec4(cameraFrameUV * 2.f - 1.f, 0.0, 1.0);
+            }
+        )"};
+
+        static constexpr char CAMERA_FRAG_SHADER[]{ R"(#version 300 es
+            #extension GL_OES_EGL_image_external_essl3 : require
+            precision mediump float;
+            in vec2 cameraFrameUV;
+            uniform samplerExternalOES cameraTexture;
+            layout(location = 0) out vec4 oFragColor;
+            void main() {
+                oFragColor = texture(cameraTexture, cameraFrameUV);
+            }
+        )"};
+
+    private:
+
+        std::string getCamId(bool frontCamera);
+
+        uint32_t width;
+        uint32_t height;
+        ACameraManager* cameraManager{};
+        ACameraDevice* cameraDevice{};
+        ACameraOutputTarget* textureTarget{};
+        ACaptureRequest* request{};
+        ANativeWindow* textureWindow{};
+        ACameraCaptureSession* textureSession{};
+        ACaptureSessionOutput* textureOutput{};
+        ACaptureSessionOutput* output{};
+        ACaptureSessionOutputContainer* outputs{};
+        GLuint cameraOESTextureId{};
+        GLuint cameraRGBATextureId{};
+        GLuint cameraShaderProgramId{};
+        GLuint clearFrameBufferId{};
+
+        android::graphics::SurfaceTexture surfaceTexture;
+        android::view::Surface surface;
+
+        constexpr GLint GetTextureUnit(GLenum texture)
+        {
+            return texture - GL_TEXTURE0;
+        }
+
+        GLuint LoadShader(GLenum shader_type, const char* shader_source)
```

Comment:
This code is the same as NativeXR/ARCore. I don't know where I can move it? New dep library?

Reply:
AndroidExtensions, maybe? I don't think we ever intend to support OpenGL for any platform other than Android.